src/System.Management.Automation/engine/parser/TypeInferenceVisitor.cs

```diff
@@ -2858,95 +2858,437 @@ private static CommandBaseAst GetPreviousPipelineCommand(CommandAst commandAst)
+                {
+                    var firstConvertExpression = attributedExpression as ConvertExpressionAst;
+                    ExpressionAst child = attributedExpression.Child;
+                    while (child is AttributedExpressionAst attributeChild)
+                    {
+                        if (firstConvertExpression is null && attributeChild is ConvertExpressionAst convertExpression)
+                        {
+                            // Multiple type constraint can be set on a variable like this: [int] [string] $Var1 = 1
+                            // But it's the left most type constraint that determines the final type.
+                            firstConvertExpression = convertExpression;
+                        }
+
+                        child = attributeChild.Child;
+                    }
+
+                    if (child is VariableExpressionAst variableExpression && AssignsToTargetVar(variableExpression))
+                    {
+                        if (firstConvertExpression is not null)
+                        {
+                            LastConstraint = firstConvertExpression.Type.TypeName;
+                        }
+                        else
+                        {
+                            SetLastAssignment(assignmentStatementAst.Right);
+                        }
+                    }
+                }
+                else if (assignmentStatementAst.Left is VariableExpressionAst variableExpression && AssignsToTargetVar(variableExpression))
+                {
+                    SetLastAssignment(assignmentStatementAst.Right);
+                }
+
+                return AstVisitAction.Continue;
             }
 
-            if (ast is CommandAst commandAst)
+            public override AstVisitAction VisitCommand(CommandAst commandAst)
             {
-                string[] variableParameters = { "PV", "PipelineVariable", "OV", "OutVariable" };
-                StaticBindingResult bindingResult = StaticParameterBinder.BindCommand(commandAst, false, variableParameters);
+                if (commandAst.Extent.StartOffset >= StopSearchOffset)
+                {
+                    return AstVisitAction.StopVisit;
+                }
+
+                string commandName = commandAst.GetCommandName();
+                if (commandName is not null && CompletionCompleters.s_varModificationCommands.Contains(commandName))
+                {
+                    StaticBindingResult bindingResult = StaticParameterBinder.BindCommand(commandAst, resolve: false, CompletionCompleters.s_varModificationParameters);
+                    if (bindingResult is not null
+                        && bindingResult.BoundParameters.TryGetValue("Name", out ParameterBindingResult variableName))
+                    {
+                        var nameValue = variableName.ConstantValue as string;
+                        if (AssignsToTargetVar(nameValue)
```

Comment:
Why not use new syntax here and below?
```suggestion
                        if (variableName.ConstantValue is string nameValue && AssignsToTargetVar(nameValue)
```

Reply:
I was under the impression that `as` would also be able to convert an int into a string if that happened to be the argument, but I can see that this is not the case.